// ngx-matomo-client/esm2022/core/tracker/internal-matomo-tracker.service.mjs
import { isPlatformBrowser } from '@angular/common';
import { inject, Injectable, NgZone, PLATFORM_ID } from '@angular/core';
import { initializeMatomoHolder } from '../holder';
import { INTERNAL_MATOMO_CONFIGURATION } from './configuration';
import * as i0 from "@angular/core";
function trimTrailingUndefinedElements(array) {
    const trimmed = [...array];
    while (trimmed.length > 0 && trimmed[trimmed.length - 1] === undefined) {
        trimmed.pop();
    }
    return trimmed;
}
export function createInternalMatomoTracker() {
    const disabled = inject(INTERNAL_MATOMO_CONFIGURATION).disabled;
    const isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    return disabled || !isBrowser ? new NoopMatomoTracker() : new InternalMatomoTracker();
}
export class InternalMatomoTracker {
    constructor() {
        this.ngZone = inject(NgZone);
        this.config = inject(INTERNAL_MATOMO_CONFIGURATION);
        initializeMatomoHolder();
    }
    /** Asynchronously call provided method name on matomo tracker instance */
    get(getter) {
        return this.pushFn(matomo => matomo[getter]());
    }
    pushFn(fn) {
        return new Promise(resolve => {
            this.push([
                function () {
                    resolve(fn(this));
                },
            ]);
        });
    }
    push(args) {
        if (this.config.runOutsideAngularZone) {
            this.ngZone.runOutsideAngular(() => {
                window._paq.push(trimTrailingUndefinedElements(args));
            });
        }
        else {
            window._paq.push(trimTrailingUndefinedElements(args));
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: InternalMatomoTracker, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: InternalMatomoTracker, providedIn: 'root', useFactory: createInternalMatomoTracker }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: InternalMatomoTracker, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                    useFactory: createInternalMatomoTracker,
                }]
        }], ctorParameters: () => [] });
export class NoopMatomoTracker {
    /** Asynchronously call provided method name on matomo tracker instance */
    async get(_) {
        return Promise.reject('MatomoTracker is disabled');
    }
    push(_) {
        // No-op
    }
    async pushFn(_) {
        return Promise.reject('MatomoTracker is disabled');
    }
}
