// ngx-matomo-client/esm2022/core/utils/script-injector.mjs
import { DOCUMENT } from '@angular/common';
import { inject, Injectable, INJECTOR, runInInjectionContext } from '@angular/core';
import { MATOMO_SCRIPT_FACTORY } from '../tracker/script-factory';
import { requireNonNull } from './coercion';
import * as i0 from "@angular/core";
export class ScriptInjector {
    constructor() {
        this.scriptFactory = inject(MATOMO_SCRIPT_FACTORY);
        this.injector = inject(INJECTOR);
        this.document = inject(DOCUMENT);
    }
    injectDOMScript(scriptUrl) {
        const scriptElement = runInInjectionContext(this.injector, () => this.scriptFactory(scriptUrl, this.document));
        const selfScript = requireNonNull(this.document.getElementsByTagName('script')[0], 'no existing script found');
        const parent = requireNonNull(selfScript.parentNode, "no script's parent node found");
        parent.insertBefore(scriptElement, selfScript);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: ScriptInjector, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: ScriptInjector, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: ScriptInjector, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });
