// ngx-matomo-client/esm2022/form-analytics/directives/track-forms.directive.mjs
import { Directive, ElementRef, inject } from '@angular/core';
import { MatomoFormAnalytics } from '../matomo-form-analytics.service';
import * as i0 from "@angular/core";
export class TrackFormsDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.tracker = inject(MatomoFormAnalytics);
    }
    ngAfterViewInit() {
        this.scan();
    }
    scan() {
        this.tracker.scanForForms(this.elementRef);
    }
    trackSubmit(element) {
        this.tracker.trackFormSubmit(element);
    }
    trackConversion(element) {
        this.tracker.trackFormConversion(element);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormsDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.2", type: TrackFormsDirective, isStandalone: true, selector: "[matomoTrackForms]", exportAs: ["matomoTrackForms"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormsDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[matomoTrackForms]',
                    standalone: true,
                    exportAs: 'matomoTrackForms',
                }]
        }] });
