// ngx-matomo-client/esm2022/form-analytics/directives/track-form-field.directive.mjs
import { booleanAttribute, Directive, ElementRef, inject, Input, } from '@angular/core';
import { throwFormNotFoundError } from './errors';
import { TrackFormDirective } from './track-form.directive';
import * as i0 from "@angular/core";
export class TrackFormFieldDirective {
    constructor() {
        this.elementRef = inject(ElementRef);
        this.form = inject(TrackFormDirective, { optional: true }) ??
            throwFormNotFoundError('[matomoTrackFormField]');
        this.initialized = false;
    }
    set matomoIgnore(ignore) {
        if (ignore) {
            this.elementRef.nativeElement.setAttribute('data-matomo-ignore', '');
        }
        else {
            this.elementRef.nativeElement.removeAttribute('data-matomo-ignore');
        }
    }
    set matomoTrackFormField(name) {
        if (name) {
            this.elementRef.nativeElement.setAttribute('data-matomo-name', name);
        }
        else {
            this.elementRef.nativeElement.removeAttribute('data-matomo-name');
        }
        if (this.initialized) {
            this.track();
        }
    }
    ngAfterViewInit() {
        this.track();
        this.initialized = true;
    }
    track() {
        this.form.track();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormFieldDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.2", type: TrackFormFieldDirective, isStandalone: true, selector: "[matomoTrackFormField]", inputs: { matomoIgnore: ["matomoIgnore", "matomoIgnore", booleanAttribute], matomoTrackFormField: "matomoTrackFormField" }, exportAs: ["matomoTrackFormField"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormFieldDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[matomoTrackFormField]',
                    standalone: true,
                    exportAs: 'matomoTrackFormField',
                }]
        }], propDecorators: { matomoIgnore: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], matomoTrackFormField: [{
                type: Input
            }] } });
