// ngx-matomo-client/esm2022/form-analytics/directives/track-form.directive.mjs
import { booleanAttribute, Directive, ElementRef, HostListener, inject, Input, } from '@angular/core';
import { MatomoFormAnalytics } from '../matomo-form-analytics.service';
import * as i0 from "@angular/core";
export class TrackFormDirective {
    set matomoIgnore(ignore) {
        if (ignore) {
            this.elementRef.nativeElement.setAttribute('data-matomo-ignore', '');
        }
        else {
            this.elementRef.nativeElement.removeAttribute('data-matomo-ignore');
        }
    }
    set matomoTrackForm(name) {
        if (name) {
            this.elementRef.nativeElement.setAttribute('data-matomo-name', name);
        }
        else {
            this.elementRef.nativeElement.removeAttribute('data-matomo-name');
        }
        if (this.initialized) {
            this.track();
        }
    }
    constructor() {
        this.elementRef = inject(ElementRef);
        this.tracker = inject(MatomoFormAnalytics);
        this.initialized = false;
        /** If true, will track a conversion after form submit */
        this.trackConversionOnSubmit = false;
        this.elementRef.nativeElement.setAttribute('data-matomo-form', '');
    }
    ngAfterViewInit() {
        this.track();
        this.initialized = true;
    }
    track() {
        this.tracker.trackForm(this.elementRef);
    }
    trackSubmit() {
        this.tracker.trackFormSubmit(this.elementRef);
    }
    trackConversion() {
        this.tracker.trackFormConversion(this.elementRef);
    }
    trackFormConversionOnSubmit() {
        if (this.trackConversionOnSubmit) {
            this.trackConversion();
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "17.3.2", type: TrackFormDirective, isStandalone: true, selector: "[matomoTrackForm]", inputs: { trackConversionOnSubmit: ["trackConversionOnSubmit", "trackConversionOnSubmit", booleanAttribute], matomoIgnore: ["matomoIgnore", "matomoIgnore", booleanAttribute], matomoTrackForm: "matomoTrackForm" }, host: { listeners: { "submit": "trackFormConversionOnSubmit()" } }, exportAs: ["matomoTrackForm"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: TrackFormDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[matomoTrackForm]',
                    standalone: true,
                    exportAs: 'matomoTrackForm',
                }]
        }], ctorParameters: () => [], propDecorators: { trackConversionOnSubmit: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], matomoIgnore: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], matomoTrackForm: [{
                type: Input
            }], trackFormConversionOnSubmit: [{
                type: HostListener,
                args: ['submit']
            }] } });
