// ngx-matomo-client/esm2022/core/matomo.module.mjs
import { NgModule } from '@angular/core';
import { MatomoOptOutFormComponent } from './directives/matomo-opt-out-form.component';
import { MatomoTrackClickDirective } from './directives/matomo-track-click.directive';
import { MatomoTrackerDirective } from './directives/matomo-tracker.directive';
import { provideMatomo, withScriptFactory } from './providers';
import * as i0 from "@angular/core";
export const MATOMO_DIRECTIVES = [
    MatomoTrackerDirective,
    MatomoTrackClickDirective,
    MatomoOptOutFormComponent,
];
function buildProviders(config, scriptFactory) {
    const features = [];
    if (scriptFactory) {
        features.push(withScriptFactory(scriptFactory));
    }
    return provideMatomo(config, ...features);
}
export class MatomoModule {
    static forRoot(config, scriptFactory) {
        return {
            ngModule: MatomoModule,
            providers: [buildProviders(config, scriptFactory)],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: MatomoModule, imports: [MatomoTrackerDirective,
            MatomoTrackClickDirective,
            MatomoOptOutFormComponent], exports: [MatomoTrackerDirective,
            MatomoTrackClickDirective,
            MatomoOptOutFormComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...MATOMO_DIRECTIVES],
                    exports: [...MATOMO_DIRECTIVES],
                }]
        }] });
/**
 * @deprecated use MatomoModule instead
 * @breaking-change 7.0.0
 */
export class NgxMatomoModule {
    static forRoot(config, scriptFactory) {
        return {
            ngModule: NgxMatomoModule,
            providers: [buildProviders(config, scriptFactory)],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoModule, imports: [MatomoModule], exports: [MatomoModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoModule, imports: [MatomoModule, MatomoModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [MatomoModule],
                    exports: [MatomoModule],
                }]
        }] });
/**
 * @deprecated use MatomoModule instead
 * @breaking-change 7.0.0
 */
export class NgxMatomoTrackerModule {
    static forRoot(config, scriptFactory) {
        return {
            ngModule: NgxMatomoTrackerModule,
            providers: [buildProviders(config, scriptFactory)],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoTrackerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoTrackerModule, imports: [NgxMatomoModule], exports: [NgxMatomoModule] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoTrackerModule, imports: [NgxMatomoModule, NgxMatomoModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoTrackerModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NgxMatomoModule],
                    exports: [NgxMatomoModule],
                }]
        }] });
