// ngx-matomo-client/esm2022/core/providers.mjs
import { ENVIRONMENT_INITIALIZER, inject, makeEnvironmentProviders, } from '@angular/core';
import { MATOMO_CONFIGURATION } from './tracker/configuration';
import { MatomoInitializerService } from './tracker/matomo-initializer.service';
import { MATOMO_SCRIPT_FACTORY } from './tracker/script-factory';
const PRIVATE_MATOMO_PROVIDERS = Symbol('MATOMO_PROVIDERS');
const PRIVATE_MATOMO_CHECKS = Symbol('MATOMO_CHECKS');
export function createMatomoFeature(kind, providers, checks) {
    return { kind, [PRIVATE_MATOMO_PROVIDERS]: providers, [PRIVATE_MATOMO_CHECKS]: checks };
}
/**
 * Return Matomo providers (typically added to an application's root module)
 *
 * Additional features may be provided as additional arguments, like this:
 * @example
 * // Simple configuration
 * providers: [ provideMatomo({ siteId: 1, trackerUrl: '...' }) ]
 *
 * // With additional features
 * providers: [
 *     provideMatomo(
 *         { siteId: 1, trackerUrl: '...' },
 *         withScriptFactory(...),
 *     )
 * ]
 *
 * // With advanced config factory function
 * providers: [
 *     provideMatomo(
 *         () => inject(MyService).getMatomoConfig(),
 *         withScriptFactory(...),
 *     )
 * ]
 *
 * @param config Matomo configuration (or configuration factory, which can use `inject`)
 * @param features Optional additional features to enable
 *
 * @see MatomoConfiguration
 * @see withScriptFactory
 * @see withRouter
 * @see withRouterInterceptors
 * @see withRouteData
 */
export function provideMatomo(config, ...features) {
    const providers = [
        {
            provide: ENVIRONMENT_INITIALIZER,
            multi: true,
            useValue() {
                inject(MatomoInitializerService).initialize();
            },
        },
    ];
    const featuresKind = [];
    if (typeof config === 'function') {
        providers.push({
            provide: MATOMO_CONFIGURATION,
            useFactory: config,
        });
    }
    else {
        providers.push({
            provide: MATOMO_CONFIGURATION,
            useValue: config,
        });
    }
    for (const feature of features) {
        providers.push(...feature[PRIVATE_MATOMO_PROVIDERS]);
        featuresKind.push(feature.kind);
    }
    for (const feature of features) {
        feature[PRIVATE_MATOMO_CHECKS]?.(featuresKind);
    }
    return makeEnvironmentProviders(providers);
}
/** Add a custom script factory to use for Matomo's script element */
export function withScriptFactory(scriptFactory) {
    return createMatomoFeature("ScriptFactory" /* CoreMatomoFeatureKind.ScriptFactory */, [
        { provide: MATOMO_SCRIPT_FACTORY, useValue: scriptFactory },
    ]);
}
