// ngx-matomo-client/esm2022/router/interceptor.mjs
import { inject, InjectionToken, INJECTOR, runInInjectionContext, } from '@angular/core';
export const MATOMO_ROUTER_INTERCEPTORS = new InjectionToken('MATOMO_ROUTER_INTERCEPTORS');
/**
 * This is not an ideal implementation, because there exist no easy way to differentiate between a class constructor and a function.
 */
function isInterceptorFn(interceptor) {
    return typeof interceptor.prototype?.beforePageTrack !== 'function';
}
class InterceptorFnAdapter {
    constructor(fn, injector) {
        this.fn = fn;
        this.injector = injector;
    }
    beforePageTrack(event) {
        return runInInjectionContext(this.injector, () => this.fn(event));
    }
}
export function provideInterceptor(typeOrFn) {
    if (isInterceptorFn(typeOrFn)) {
        return {
            provide: MATOMO_ROUTER_INTERCEPTORS,
            multi: true,
            useFactory: () => new InterceptorFnAdapter(typeOrFn, inject(INJECTOR)),
        };
    }
    else {
        return {
            provide: MATOMO_ROUTER_INTERCEPTORS,
            multi: true,
            useClass: typeOrFn,
        };
    }
}
export function provideInterceptors(types) {
    if (!types) {
        return [];
    }
    return types.map(provideInterceptor);
}
