// ngx-matomo-client/esm2022/router/matomo-router.module.mjs
import { forwardRef, Inject, NgModule, Optional, SkipSelf, } from '@angular/core';
import { ɵMATOMO_ROUTER_ENABLED as MATOMO_ROUTER_ENABLED } from 'ngx-matomo-client/core';
import { MATOMO_ROUTER_CONFIGURATION, } from './configuration';
import { provideInterceptors } from './interceptor';
import * as i0 from "@angular/core";
import * as i1 from "./matomo-router.service";
export class MatomoRouterModule {
    constructor(router, parent, parentDeprecated) {
        this.router = router;
        if (!parent && !parentDeprecated) {
            // Do not initialize if it is already (by a parent module)
            this.router.initialize();
        }
    }
    static forRoot(configWithInterceptors = {}) {
        // Note: not using "rest" syntax here, in order to avoid any dependency on tslib (and reduce package size)
        // The only drawback of this is that MATOMO_ROUTER_CONFIGURATION will actually
        // contain a reference to provided interceptors
        return {
            ngModule: MatomoRouterModule,
            providers: [
                { provide: MATOMO_ROUTER_CONFIGURATION, useValue: configWithInterceptors },
                provideInterceptors(configWithInterceptors.interceptors),
            ],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoRouterModule, deps: [{ token: i1.MatomoRouter }, { token: MatomoRouterModule, optional: true, skipSelf: true }, { token: forwardRef(() => NgxMatomoRouterModule), optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: MatomoRouterModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoRouterModule, providers: [{ provide: MATOMO_ROUTER_ENABLED, useValue: true }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoRouterModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [{ provide: MATOMO_ROUTER_ENABLED, useValue: true }],
                }]
        }], ctorParameters: () => [{ type: i1.MatomoRouter }, { type: MatomoRouterModule, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }] }, { type: NgxMatomoRouterModule, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => NgxMatomoRouterModule)]
                }, {
                    type: Optional
                }, {
                    type: SkipSelf
                }] }] });
/**
 * @deprecated use MatomoRouterModule instead
 * @breaking-change 7.0.0
 */
export class NgxMatomoRouterModule {
    constructor(router, parent, parentDeprecated) {
        this.router = router;
        if (!parent && !parentDeprecated) {
            // Do not initialize if it is already (by a parent module)
            this.router.initialize();
        }
    }
    static forRoot(configWithInterceptors = {}) {
        // Note: not using "rest" syntax here, in order to avoid any dependency on tslib (and reduce package size)
        // The only drawback of this is that MATOMO_ROUTER_CONFIGURATION will actually
        // contain a reference to provided interceptors
        return {
            ngModule: NgxMatomoRouterModule,
            providers: [
                { provide: MATOMO_ROUTER_CONFIGURATION, useValue: configWithInterceptors },
                provideInterceptors(configWithInterceptors.interceptors),
            ],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoRouterModule, deps: [{ token: i1.MatomoRouter }, { token: MatomoRouterModule, optional: true, skipSelf: true }, { token: forwardRef(() => NgxMatomoRouterModule), optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoRouterModule }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoRouterModule, providers: [{ provide: MATOMO_ROUTER_ENABLED, useValue: true }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: NgxMatomoRouterModule, decorators: [{
            type: NgModule,
            args: [{
                    providers: [{ provide: MATOMO_ROUTER_ENABLED, useValue: true }],
                }]
        }], ctorParameters: () => [{ type: i1.MatomoRouter }, { type: MatomoRouterModule, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }] }, { type: NgxMatomoRouterModule, decorators: [{
                    type: Inject,
                    args: [forwardRef(() => NgxMatomoRouterModule)]
                }, {
                    type: Optional
                }, {
                    type: SkipSelf
                }] }] });
