// ngx-matomo-client/esm2022/router/page-url-provider.mjs
import { APP_BASE_HREF, LocationStrategy } from '@angular/common';
import { inject, InjectionToken } from '@angular/core';
import { of } from 'rxjs';
import { INTERNAL_ROUTER_CONFIGURATION } from './configuration';
export const MATOMO_PAGE_URL_PROVIDER = new InjectionToken('MATOMO_PAGE_URL_PROVIDER', {
    factory: () => new DefaultPageUrlProvider(inject(INTERNAL_ROUTER_CONFIGURATION), inject(APP_BASE_HREF, { optional: true }), inject(LocationStrategy)),
});
function trimTrailingSlash(str) {
    return str.endsWith('/') ? str.slice(0, -1) : str;
}
export class DefaultPageUrlProvider {
    constructor(config, baseHref, locationStrategy) {
        this.config = config;
        this.baseHref = baseHref;
        this.locationStrategy = locationStrategy;
    }
    getCurrentPageUrl(event) {
        const url = this.config.prependBaseHref
            ? this.getBaseHrefWithoutTrailingSlash() + event.urlAfterRedirects
            : event.urlAfterRedirects;
        return of(url);
    }
    getBaseHrefWithoutTrailingSlash() {
        return trimTrailingSlash(this.baseHref ?? this.locationStrategy.getBaseHref() ?? '');
    }
}
