// ngx-matomo-client/esm2022/router/providers.mjs
import { ENVIRONMENT_INITIALIZER, inject } from '@angular/core';
import { ɵcreateMatomoFeature as createMatomoFeature, ɵMATOMO_ROUTER_ENABLED as MATOMO_ROUTER_ENABLED, } from 'ngx-matomo-client/core';
import { MATOMO_ROUTER_CONFIGURATION } from './configuration';
import { provideInterceptor, provideInterceptors, } from './interceptor';
import { MATOMO_ROUTE_DATA_KEY, MatomoRouteDataInterceptor, } from './interceptors/route-data-interceptor';
import { MatomoRouter } from './matomo-router.service';
/** Enable automatic page views tracking */
export function withRouter(config) {
    const providers = [
        { provide: MATOMO_ROUTER_ENABLED, useValue: true },
        { provide: MATOMO_ROUTER_CONFIGURATION, useValue: config },
        {
            provide: ENVIRONMENT_INITIALIZER,
            multi: true,
            useValue() {
                inject(MatomoRouter).initialize();
            },
        },
    ];
    return createMatomoFeature("Router" /* RouterMatomoFeatureKind.Router */, providers);
}
/** Add some matomo router interceptors */
export function withRouterInterceptors(interceptors) {
    return createMatomoFeature("RouterInterceptors" /* RouterMatomoFeatureKind.RouterInterceptors */, provideInterceptors(interceptors), featuresKind => {
        if (!featuresKind.includes("Router" /* RouterMatomoFeatureKind.Router */)) {
            throw new Error(`Matomo feature withRouterInterceptors() cannot be used without router feature! Did you forget to call withRouter()?`);
        }
    });
}
/**
 * Enable retrieval of tracking information from route data
 *
 * @see MatomoRouteData
 * @param key A custom key to get lookup route data - default is 'matomo'
 */
export function withRouteData(key) {
    const providers = [provideInterceptor(MatomoRouteDataInterceptor)];
    if (key) {
        providers.push({ provide: MATOMO_ROUTE_DATA_KEY, useValue: key });
    }
    return createMatomoFeature("BuiltInRouteDataInterceptor" /* RouterMatomoFeatureKind.BuiltInRouteDataInterceptor */, providers, featuresKind => {
        if (!featuresKind.includes("Router" /* RouterMatomoFeatureKind.Router */)) {
            throw new Error(`Matomo feature withRouteData() cannot be used without router feature! Did you forget to call withRouter()?`);
        }
    });
}
