// ngx-matomo-client/esm2022/testing/matomo-testing.module.mjs
import { NgModule } from '@angular/core';
import { MATOMO_DIRECTIVES } from 'ngx-matomo-client/core';
import { provideMatomoTesting } from './providers';
import * as i0 from "@angular/core";
import * as i1 from "ngx-matomo-client/core";
export class MatomoTestingModule {
    static forRoot() {
        return {
            ngModule: MatomoTestingModule,
            providers: [provideMatomoTesting()],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingModule, imports: [i1.MatomoTrackerDirective, i1.MatomoTrackClickDirective, i1.MatomoOptOutFormComponent], exports: [i1.MatomoTrackerDirective, i1.MatomoTrackClickDirective, i1.MatomoOptOutFormComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingModule, imports: [i1.MatomoOptOutFormComponent] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...MATOMO_DIRECTIVES],
                    exports: [...MATOMO_DIRECTIVES],
                }]
        }] });
