// ngx-matomo-client/esm2022/testing/matomo-testing-instance.mjs
/**
 * No-op implementation of {@link MatomoInstance}
 */
export class MatomoTestingInstance {
    areCookiesEnabled() {
        return false;
    }
    getAttributionCampaignKeyword() {
        return '';
    }
    getAttributionCampaignName() {
        return '';
    }
    getAttributionInfo() {
        return [];
    }
    getAttributionReferrerTimestamp() {
        return '';
    }
    getAttributionReferrerUrl() {
        return '';
    }
    getCrossDomainLinkingUrlParameter() {
        return '';
    }
    getCurrentUrl() {
        return '';
    }
    getCustomDimension(_customDimensionId) {
        return '';
    }
    getCustomPagePerformanceTiming() {
        return '';
    }
    getCustomVariable(_index, _scope) {
        return '';
    }
    getEcommerceItems() {
        return [];
    }
    getExcludedReferrers() {
        return [];
    }
    getLinkTrackingTimer() {
        return 0;
    }
    getMatomoUrl() {
        return '';
    }
    getPageViewId() {
        return '';
    }
    getCustomData() {
        return undefined;
    }
    getPiwikUrl() {
        return '';
    }
    getRememberedConsent() {
        return '';
    }
    getRememberedCookieConsent() {
        return '';
    }
    getUserId() {
        return '';
    }
    getVisitorId() {
        return '';
    }
    getVisitorInfo() {
        return [];
    }
    hasCookies() {
        return false;
    }
    hasRememberedConsent() {
        return false;
    }
    isConsentRequired() {
        return false;
    }
    isUserOptedOut() {
        return false;
    }
}
