// ngx-matomo-client/esm2022/testing/matomo-testing-tracker.mjs
import { Injectable } from '@angular/core';
import { MatomoTracker } from 'ngx-matomo-client/core';
import { MatomoTestingInstance } from './matomo-testing-instance';
import * as i0 from "@angular/core";
/**
 * No-op implementation of {@link MatomoTracker}
 *
 * All commands are memoized and can later be retrieved using
 * {@link getAllCommands getAllCommands()} or {@link getCommand getCommand(index)}.
 *
 * All <i>getter</i> methods will immediately resolve to an <i>empty value</i>.
 * This can be customized by setting a custom Matomo instance with {@link setMatomoInstance setMatomoInstance()}.
 */
export class MatomoTestingTracker extends MatomoTracker {
    #fakeInstance = new MatomoTestingInstance();
    #paq = [];
    /** Retrieve the current Matomo instance */
    getMatomoInstance() {
        return this.#fakeInstance;
    }
    /** Set the current matomo instance */
    setMatomoInstance(instance) {
        this.#fakeInstance = instance;
    }
    /** Retrieve all memoized commands */
    getAllCommands() {
        return [...this.#paq];
    }
    /** Retrieve nth memoized command */
    getCommand(index) {
        return this.#paq[index];
    }
    /** Clear all memoized commands */
    reset() {
        this.#paq = [];
    }
    push(command) {
        this.#paq.push(command);
    }
    async pushFn(fn) {
        return fn(this.#fakeInstance);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingTracker, deps: null, target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingTracker }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTestingTracker, decorators: [{
            type: Injectable
        }] });
