// ngx-matomo-client/esm2022/form-analytics/matomo-form-analytics.service.mjs
import { inject, Injectable } from '@angular/core';
import { ɵInternalMatomoTracker as InternalMatomoTracker } from 'ngx-matomo-client/core';
import { coerceElement } from './utils/coercion';
import * as i0 from "@angular/core";
export class MatomoFormAnalytics {
    constructor() {
        this.delegate = inject(InternalMatomoTracker);
    }
    disableFormAnalytics() {
        this.delegate.push(['FormAnalytics::disableFormAnalytics']);
    }
    enableFormAnalytics() {
        this.delegate.push(['FormAnalytics::enableFormAnalytics']);
    }
    enableDebugMode() {
        this.delegate.push(['FormAnalytics::enableDebugMode']);
    }
    setTrackingTimer(delayInMilliSeconds) {
        this.delegate.push(['FormAnalytics::setTrackingTimer', delayInMilliSeconds]);
    }
    scanForForms(root) {
        this.delegate.push(['FormAnalytics::scanForForms', coerceElement(root)]);
    }
    trackForm(formElement) {
        this.delegate.push(['FormAnalytics::trackForm', coerceElement(formElement)]);
    }
    trackFormSubmit(formElement) {
        this.delegate.push(['FormAnalytics::trackFormSubmit', coerceElement(formElement)]);
    }
    trackFormConversion(formElementOrName, formId) {
        this.delegate.push([
            'FormAnalytics::trackFormConversion',
            typeof formElementOrName === 'string' ? formElementOrName : coerceElement(formElementOrName),
            formId,
        ]);
    }
    isFormAnalyticsEnabled() {
        return this.delegate.get('FormAnalytics::isFormAnalyticsEnabled');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalytics, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalytics, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalytics, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
