// ngx-matomo-client/esm2022/form-analytics/matomo-form-analytics-initializer.service.mjs
import { isPlatformBrowser } from '@angular/common';
import { inject, Injectable, PLATFORM_ID } from '@angular/core';
import { MatomoTracker, ɵappendTrailingSlash as appendTrailingSlash, ɵASYNC_INTERNAL_MATOMO_CONFIGURATION as ASYNC_INTERNAL_MATOMO_CONFIGURATION, ɵgetTrackersConfiguration as getTrackersConfiguration, ɵisAutoConfigurationMode as isAutoConfigurationMode, ɵisExplicitTrackerConfiguration as isExplicitTrackerConfiguration, ɵrunOnce as runOnce, ɵScriptInjector as ScriptInjector, } from 'ngx-matomo-client/core';
import { debounceTime, identity } from 'rxjs';
import { INTERNAL_MATOMO_FORM_ANALYTICS_CONFIGURATION } from './configuration';
import { MatomoFormAnalytics } from './matomo-form-analytics.service';
import * as i0 from "@angular/core";
const DEFAULT_SCRIPT_SUFFIX = 'plugins/FormAnalytics/tracker.min.js';
export class MatomoFormAnalyticsInitializer {
    constructor() {
        this.config = inject(INTERNAL_MATOMO_FORM_ANALYTICS_CONFIGURATION);
        this.coreConfig = inject(ASYNC_INTERNAL_MATOMO_CONFIGURATION);
        this.scriptInjector = inject(ScriptInjector);
        this.tracker = inject(MatomoTracker);
        this.formAnalytics = inject(MatomoFormAnalytics);
        this.platformId = inject(PLATFORM_ID);
        this.initialize = runOnce(async () => {
            // Do not set-up router if running on server
            if (!isPlatformBrowser(this.platformId)) {
                return;
            }
            if (this.config.disabled) {
                this.formAnalytics.disableFormAnalytics();
                return;
            }
            if (this.config.loadScript) {
                const scriptUrl = typeof this.config.loadScript === 'boolean'
                    ? await this.buildDefaultScriptUrl()
                    : this.config.loadScript;
                this.scriptInjector.injectDOMScript(scriptUrl);
            }
            if (this.config.autoScan) {
                const delayOp = this.config.autoScanDelay
                    ? debounceTime(this.config.autoScanDelay)
                    : identity;
                this.pageTrackedSubscription = this.tracker.pageViewTracked
                    .pipe(delayOp)
                    .subscribe(() => this.formAnalytics.scanForForms());
            }
        });
    }
    ngOnDestroy() {
        this.pageTrackedSubscription?.unsubscribe();
    }
    async buildDefaultScriptUrl() {
        const config = await this.coreConfig;
        if (isAutoConfigurationMode(config) && isExplicitTrackerConfiguration(config)) {
            const [mainTracker] = getTrackersConfiguration(config);
            if (mainTracker) {
                return appendTrailingSlash(mainTracker.trackerUrl) + DEFAULT_SCRIPT_SUFFIX;
            }
        }
        throw new Error('Cannot resolve default matomo FormAnalytics plugin script url. ' +
            'Please explicitly provide `loadScript` configuration property instead of `true`');
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsInitializer, deps: [], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsInitializer, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsInitializer, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
