// ngx-matomo-client/esm2022/form-analytics/matomo-form-analytics.module.mjs
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { MATOMO_FORM_ANALYTICS_CONFIGURATION, } from './configuration';
import { TrackFormFieldDirective } from './directives/track-form-field.directive';
import { TrackFormDirective } from './directives/track-form.directive';
import { TrackFormsDirective } from './directives/track-forms.directive';
import * as i0 from "@angular/core";
import * as i1 from "./matomo-form-analytics-initializer.service";
export const MATOMO_FORM_ANALYTICS_DIRECTIVES = [
    TrackFormDirective,
    TrackFormsDirective,
    TrackFormFieldDirective,
];
export class MatomoFormAnalyticsModule {
    constructor(formAnalytics, parent) {
        this.formAnalytics = formAnalytics;
        if (!parent) {
            // Do not initialize if it is already (by a parent module)
            this.formAnalytics.initialize();
        }
    }
    static forRoot(config = {}) {
        return {
            ngModule: MatomoFormAnalyticsModule,
            providers: [{ provide: MATOMO_FORM_ANALYTICS_CONFIGURATION, useValue: config }],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsModule, deps: [{ token: i1.MatomoFormAnalyticsInitializer }, { token: MatomoFormAnalyticsModule, optional: true, skipSelf: true }], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsModule, imports: [TrackFormDirective,
            TrackFormsDirective,
            TrackFormFieldDirective], exports: [TrackFormDirective,
            TrackFormsDirective,
            TrackFormFieldDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoFormAnalyticsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [...MATOMO_FORM_ANALYTICS_DIRECTIVES],
                    exports: [...MATOMO_FORM_ANALYTICS_DIRECTIVES],
                }]
        }], ctorParameters: () => [{ type: i1.MatomoFormAnalyticsInitializer }, { type: MatomoFormAnalyticsModule, decorators: [{
                    type: Optional
                }, {
                    type: SkipSelf
                }] }] });
