// ngx-matomo-client/esm2022/core/directives/matomo-tracker.directive.mjs
import { Directive, Input } from '@angular/core';
import { fromEvent, merge } from 'rxjs';
import { requireNonNull } from '../utils/coercion';
import * as i0 from "@angular/core";
import * as i1 from "../tracker/matomo-tracker.service";
function coerceEventNames(input) {
    if (input && input.length > 0) {
        return Array.isArray(input) ? input : [input];
    }
    else {
        return undefined;
    }
}
export class MatomoTrackerDirective {
    constructor(tracker, elementRef) {
        this.tracker = tracker;
        this.elementRef = elementRef;
    }
    /** Track a Matomo event whenever specified DOM event is triggered */
    set matomoTracker(input) {
        const eventNames = coerceEventNames(input);
        this.sub?.unsubscribe();
        if (eventNames) {
            const handlers = eventNames.map(eventName => fromEvent(this.elementRef.nativeElement, eventName));
            this.sub = merge(...handlers).subscribe(() => this.trackEvent());
        }
        else {
            this.sub = undefined;
        }
    }
    ngOnDestroy() {
        this.sub?.unsubscribe();
    }
    trackEvent(arg1, arg2) {
        let category = this.matomoCategory;
        let action = this.matomoAction;
        let name = this.matomoName;
        let value = this.matomoValue;
        if (typeof arg1 === 'object') {
            category = arg1.category ?? category;
            action = arg1.action ?? action;
            name = arg1.name ?? name;
            value = arg1.value ?? value;
        }
        else if (typeof arg1 === 'string') {
            name = arg1;
            if (typeof arg2 === 'number') {
                value = arg2;
            }
        }
        else if (typeof arg1 === 'number') {
            value = arg1;
        }
        this.tracker.trackEvent(requireNonNull(category, 'matomo category is required'), requireNonNull(action, 'matomo action is required'), name, value);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTrackerDirective, deps: [{ token: i1.MatomoTracker }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.3.2", type: MatomoTrackerDirective, isStandalone: true, selector: "[matomoTracker]", inputs: { matomoCategory: "matomoCategory", matomoAction: "matomoAction", matomoName: "matomoName", matomoValue: "matomoValue", matomoTracker: "matomoTracker" }, exportAs: ["matomo"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.3.2", ngImport: i0, type: MatomoTrackerDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[matomoTracker]',
                    exportAs: 'matomo',
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: i1.MatomoTracker }, { type: i0.ElementRef }], propDecorators: { matomoCategory: [{
                type: Input
            }], matomoAction: [{
                type: Input
            }], matomoName: [{
                type: Input
            }], matomoValue: [{
                type: Input
            }], matomoTracker: [{
                type: Input
            }] } });
